tsdb/shard.go

```diff
@@ -761,17 +821,18 @@ func (s *Shard) WriteTo(w io.Writer) (int64, error) {
 
 // CreateIterator returns an iterator for the data in the shard.
 func (s *Shard) CreateIterator(measurement string, opt query.IteratorOptions) (query.Iterator, error) {
+	s.mu.RLock()
```

Comment:
Can you grab a reference to the engine and use that instead of holding a read lock for the whole func?  The locks were lower to avoid lock contention when snapshotting the cache.